awscli/autoprompt/widgets.py

```diff
@@ -0,0 +1,208 @@
+# Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+from prompt_toolkit.formatted_text import HTML, to_formatted_text
+from prompt_toolkit.formatted_text.utils import fragment_list_to_text
+from prompt_toolkit.buffer import Buffer
+from prompt_toolkit.document import Document
+from prompt_toolkit.layout import HSplit, Window, VSplit
+from prompt_toolkit.layout.controls import BufferControl
+from prompt_toolkit.layout.dimension import Dimension
+from prompt_toolkit.layout.layout import ConditionalContainer
+from prompt_toolkit.layout.processors import Processor, Transformation
+from prompt_toolkit.widgets import Frame, HorizontalLine
+
+from awscli.autoprompt.filters import (
+    help_section_visible, doc_window_has_focus, search_input_has_focus,
+    input_buffer_has_focus, is_history_mode
+)
+
+
+class FormatTextProcessor(Processor):
+    """This Processor is used to transform formatted text into a useable
+    format inside a ``prompt_toolkit.buffer.Buffer``.
+
+    """
+    def apply_transformation(self, text_input):
+        # https://python-prompt-toolkit.readthedocs.io/en/master/pages/reference.html#module-prompt_toolkit.formatted_text
+        fragments = to_formatted_text(
+            HTML(fragment_list_to_text(text_input.fragments)))
+        return Transformation(fragments)
+
+
+class HelpPanelWidget:
+    DIMENSIONS = {
+        'help_window_width_max': 35,
+        'help_window_width_pref': 25,
+    }
+    style = '<style fg="darkturquoise">'
+    eol = '\n'
+
+    def __init__(self):
+        help_doc_buffer = self.create_help_buffer(
+            'help_doc', self.doc_window_key_binding_text()
+        )
+        self.help_doc_window = self.create_help_window(
+            help_doc_buffer, title='Doc panel help')
+        help_input_buffer = self.create_help_buffer(
+            'help_input', self.input_buffer_key_binding_text()
+        )
+        self.help_input_window = self.create_help_window(
+            help_input_buffer, title='Prompt panel help')
+        self.container = HSplit([
+            ConditionalContainer(
+                self.help_doc_window,
+                help_section_visible & (doc_window_has_focus
+                                        | search_input_has_focus)),
+            ConditionalContainer(
+                self.help_input_window,
+                help_section_visible & input_buffer_has_focus)
+        ])
```

Comment:
Instead of chaining the `help_section_visible` filter for both of these ConditionalContainer, could we just encapsulate the `HSplit` with a `ConditionalContainer` and hoist the `help_section_visible` to that?